sdk/helper/testcluster/docker/environment.go

```diff
@@ -0,0 +1,967 @@
+		ClientCAs:      n.Cluster.RootCAs,
+		ClientAuth:     tls.RequestClientCert,
+		NextProtos:     []string{"h2", "http/1.1"},
+		GetCertificate: certGetter.GetCertificate,
+	}
+
+	n.tlsConfig = tlsConfig
+
+	err = os.WriteFile(filepath.Join(n.WorkDir, "ca.pem"), n.Cluster.CACertPEM, 0o755)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func NewTestDockerCluster(t *testing.T, opts *DockerClusterOptions) *DockerCluster {
+	if opts == nil {
+		opts = &DockerClusterOptions{}
+	}
+	if opts.ClusterName == "" {
+		opts.ClusterName = strings.ReplaceAll(t.Name(), "/", "-")
+	}
+	if opts.Logger == nil {
+		opts.Logger = logging.NewVaultLogger(log.Trace).Named(t.Name())
+	}
+	if opts.NetworkName == "" {
+		opts.NetworkName = os.Getenv("TEST_DOCKER_NETWORK_NAME")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
+	t.Cleanup(cancel)
+
+	dc, err := NewDockerCluster(ctx, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return dc
+}
+
+func NewDockerCluster(ctx context.Context, opts *DockerClusterOptions) (*DockerCluster, error) {
+	api, err := dockhelper.NewDockerAPI()
+	if err != nil {
+		return nil, err
+	}
+
+	if opts == nil {
+		opts = &DockerClusterOptions{}
+	}
+	if opts.Logger == nil {
+		opts.Logger = log.NewNullLogger()
+	}
+
+	if opts.VaultBinary != "" {
+		f, err := os.Open(opts.VaultBinary)
+		if err != nil {
+			return nil, err
+		}
+		data, err := io.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+		bCtx := dockhelper.NewBuildContext()
+		bCtx["vault"] = &dockhelper.FileContents{
+			Data: data,
+			Mode: 0o755,
+		}
+
+		containerFile := fmt.Sprintf(`
+	FROM %s:%s
+	COPY vault /bin/vault
+`, opts.ImageRepo, opts.ImageTag)
+
+		// TODO would be nice to use current SHA as tag
+		newTag := opts.ImageTag + "-testing"
+		_, err = dockhelper.BuildImage(ctx, api, containerFile, bCtx,
+			dockhelper.BuildRemove(true), dockhelper.BuildForceRemove(true),
+			dockhelper.BuildPullParent(true),
+			dockhelper.BuildTags([]string{opts.ImageRepo + ":" + newTag}))
+		if err != nil {
+			return nil, err
+		}
+		newOpts := *opts
+		newOpts.ImageTag = newTag
+		opts = &newOpts
+	}
+
+	dc := &DockerCluster{
+		dockerAPI:   api,
+		RaftStorage: true,
+		ClusterName: opts.ClusterName,
+		Logger:      opts.Logger,
+	}
+
+	if err := dc.setupDockerCluster(ctx, opts); err != nil {
+		dc.Cleanup()
+		return nil, err
+	}
+
+	return dc, nil
+}
+
+// DockerClusterNode represents a single instance of Vault in a cluster
+type DockerClusterNode struct {
+	NodeID               string
+	HostPort             string
+	client               *api.Client
+	ServerCert           *x509.Certificate
+	ServerCertBytes      []byte
+	ServerCertPEM        []byte
+	ServerCertPEMFile    string
+	ServerKey            *ecdsa.PrivateKey
+	ServerKeyPEM         []byte
+	ServerKeyPEMFile     string
+	tlsConfig            *tls.Config
+	WorkDir              string
+	Cluster              *DockerCluster
+	container            *types.ContainerJSON
```

Comment:
Looks like I was mistaken: `NodeID` is the name of this node, but `container.ID` is the docker container ID we'd need to use for poking the container.

Any chance we could get this exported?

This was the last item that prevents ACME testing, changing it locally seems to give me a working test suite :partying_face:

Reply:
Done.

Reply:
Thank you!!